extensions/Worker.Extensions.Http.AspNetCore/src/Coordinator/ContextReference.cs

```diff
@@ -0,0 +1,81 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+
+namespace Microsoft.Azure.Functions.Worker.Extensions.Http.AspNetCore.Coordinator
+{
+    internal class ContextReference : IDisposable
+    {
+        private readonly TaskCompletionSource<bool> _functionStartTask = new();
+        private readonly TaskCompletionSource<bool> _functionCompletionTask = new();
+
+        private TaskCompletionSource<HttpContext> _httpContextValueSource = new();
+        private TaskCompletionSource<FunctionContext> _functionContextValueSource = new();
+
+        private CancellationToken _token;
+        private CancellationTokenRegistration _tokenRegistration;
+
+        public ContextReference(string invocationId)
+        {
+            InvocationId = invocationId;
+        }
+
+        public string InvocationId { get; private set; }
+
+        public TaskCompletionSource<bool> FunctionStartTask { get => _functionStartTask; }
+
+        public TaskCompletionSource<bool> FunctionCompletionTask { get => _functionCompletionTask; }
+
+        public TaskCompletionSource<HttpContext> HttpContextValueSource { get => _httpContextValueSource; set => _httpContextValueSource = value; }
+
+        public TaskCompletionSource<FunctionContext> FunctionContextValueSource { get => _functionContextValueSource; set => _functionContextValueSource = value; }
+
+        internal void SetCancellationToken(CancellationToken token)
+        {
+            _token = token;
+            _tokenRegistration = _token.Register(() =>
+            {
+                _functionStartTask.TrySetCanceled();
+                _functionCompletionTask.TrySetCanceled();
+                _functionContextValueSource.TrySetCanceled();
+                _httpContextValueSource.TrySetCanceled();
+            });
+        }
+
+        internal void StartFunction()
```

Comment:
Can we make this an awaitable/`Task` returning method and return the `FunctionCompletionTask.Task` to simplify consumption?

Reply:
The caller of this shouldn't have to await anything though -- it's the consumer of the FunctionStartTask that is waiting. This call signals that "anyone awaiting the start of the function can now continue"